src/tree/AATree.scala

```diff
@@ -0,0 +1,220 @@
+/**
+ * This file is part of Scalacaster project,
+ * https://github.com/vkostyukov/scalacaster and written by Amanj Sherwany,
+ * http://www.amanj.me
+ *
+ * AA tree http://en.wikipedia.org/wiki/AA_tree
+ *
+ * Performance: (From Wikipedia)
+ * The performance of an AA tree is equivalent to the performance of a
+ * red-black tree. While an AA tree makes more rotations than a red-black tree,
+ * the simpler algorithms tend to be faster, and all of this balances out to
+ * result in similar performance. A red-black tree is more consistent in its
+ * performance than an AA tree, but an AA tree tends to be flatter, which
+ * results in slightly faster search times
+ *
+ * Insert - O(log n)
+ * Lookup - O(log n)  
+ * Remove - O(log n)
+ */
+
+object AATree {
+
+  sealed abstract class AATree[+T: Ordering] {
+    val value: T
+    val level: Int
+    val left: AATree[T]
+    val right: AATree[T]
+
+    def member[B >: T: Ordering](a: B): Boolean = this match {
+      case Empty                                                => false
+      case Fork(v, _, l, r) if a == v                           => true
+      case Fork(v, _, l, r) if implicitly[Ordering[B]].lt(a, v) => l.member(v)
+      case Fork(v, _, l, r)                                     => r.member(v)
+    }
+
+    def +[B >: T: Ordering](a: B): AATree[B] = this match {
+      case Empty                                                 =>
+        Fork(a, 1, Empty, Empty)
+      case Fork(v, lv, l, r) if implicitly[Ordering[B]].lt(a, v) => 
+        Fork(v, lv, l.insert(a), r).rebalanceTree
+      case Fork(v, lv, l, r) if implicitly[Ordering[B]].gt(a, v) => 
+        Fork(v, lv, l, r.insert(a)).rebalanceTree
+      case _                                                     =>
+        this
+    }
+
+    def insert[B >: T: Ordering](a: B): AATree[B] = insert(a)
+    def add[B >: T: Ordering](a: B): AATree[B] = insert(a)
+
+    def insertAll[B >: T: Ordering](as: List[B]): AATree[B] = {
```

Comment:
I needed this to do merge, but I changed my merge function now not to rely on insert all